### ISSUE TITLE
rpc can only return `type` not `interface`/`class`

### ISSUE THREAD

**Issue by helloimalastair** [cloudflare/workerd#2003]
[Repro](https://github.com/helloimalastair/rpc-collapse)

When calling a DO RPC method, return type collapses to `Promise<undefined>`/`never` if using an `interface`. When using a `type`, return values are as expected.

**Comment by helloimalastair** [comment]
Just ran into it not working with `WebSocket`s either, which are `class`es, though not sure you can move a `WebSocket` anyway...

**Comment by kentonv** [comment]
As discussed, it's expected (and documented) that class instances are only supported if they extend `RpcTarget` or are instances of one of the specific API types that are documented to work. `WebSocket` is not currently supported.

That said, it sounds like there is arguably a bug in the TypeScript types. It sounds like the types treat `interface` as strictly referring to class instances, but actually an `interface` can be satisfied by a plain object too.

@mrbbot Is there any way for us to more specifically filter out class instances using the TypeScript type system?

**Comment by kentonv** [comment]
Err, apparently @mrbbot recently left the company. Sorry for the spurious ping @mrbbot, hope things go well at whatever's next for you.

I'm not sure who that leaves as the typescript wizard now. @GregBrimble any thoughts?

**Comment by katis** [comment]
Recursively mapping the return type to a type alias helped me work around this issue with some 3rd party types

```typescript
type InterfaceToType<T> = T extends object ? { [P in keyof T]: InterfaceToType<T[P]> } : T

class Foo extends RpcTarget {
	async bar(): Promise<InterfaceToType<InterfaceOfThirdParty>> {
            // ...
        }
}
```

**Comment by GregBrimble** [comment]
@penalosa would be my next go-to!

**Comment by ngbrown** [comment]
I had a hell of a time trying to get typings working between my Worker RPC service class extending `WorkerEntrypoint<Env>` and the consuming Pages function.

They are setup as two separate components in a mono repo with separate `tsconfig.json` files. I tried using `composite` and `references` in my `tsconfig.json` files. I tried not using them. Nothing seemed to work. The branded types seem to have a really hard time transferring across TypeScript contexts. If the component with the worker providing the service emits typescript definitions `emitDeclarationOnly`, the output includes a copy of the class `WorkerEntrypoint`, without the branding. So maybe that is what is happening in memory when TypeScript analyzes the whole solution.

I also tried putting `/// <reference types="@cloudflare/workers-types" preserve="true" />` at the top of various files. That also didn't help.

The output types the functions use are all `type` definitions, after I converted from `interface`... I saw that is another pull request.

Anyways, in the end, I just wanted my types to resolve to something besides `never` and `null`. I also wanted the `[Symbol.dispose]` on them to be correctly in place. So I finally patched the `@cloudflare/workers-types` package to remove the `never` option and that worked:

```patch
diff --git a/node_modules/@cloudflare/workers-types/2023-07-01/index.ts b/node_modules/@cloudflare/workers-types/2023-07-01/index.ts
index 19d049f..8b653d8 100644
--- a/node_modules/@cloudflare/workers-types/2023-07-01/index.ts
+++ b/node_modules/@cloudflare/workers-types/2023-07-01/index.ts
@@ -5000,9 +5000,7 @@ export declare namespace Rpc {
   // Intersecting with `(Maybe)Provider` allows pipelining.
   type Result<R> = R extends Stubable
     ? Promise<Stub<R>> & Provider<R>
-    : R extends Serializable
-      ? Promise<Stubify<R> & MaybeDisposable<R>> & MaybeProvider<R>
-      : never;
+    : Promise<Stubify<R> & MaybeDisposable<R>> & MaybeProvider<R>;
   // Type for method or property on an RPC interface.
   // For methods, unwrap `Stub`s in parameters, and rewrite returns to be `Result`s.
   // Unwrapping `Stub`s allows calling with `Stubable` arguments.
```

If you are wondering why it's a dated version of the types, I tried without the date path and with the date path, and neither had any difference. The templates default to no date, but the `@cloudflare/workers-types` readme says that the types are then the oldest version. I'm not sure what's up with that.

I don't think it's worth it to shut down the typing system by forcing a `never` output type if all the stars don't align perfectly. TypeScript works best as a hinting system, not as a fool-proof compiler check.

**Comment by jfsiii** [comment]
I [reached out](https://twitter.com/JFSIII/status/1804246302932300160) to @andarist [for help on Twitter](https://twitter.com/JFSIII/status/1804246302932300160). Here's a paste of the thread:

-----
[Mateusz Burzyński](https://twitter.com/AndaristRake) [@AndaristRake](https://twitter.com/AndaristRake) [Jun 22](https://twitter.com/AndaristRake/status/1804411002164838447)
is the problem here with `bazBumOptional` at the bottom?

[https://tsplay.dev/mqgJQw](https://t.co/eQCQEKuKdb) 

[John Schulz](https://twitter.com/JFSIII) [@JFSIII](https://twitter.com/JFSIII) [Jun 22](https://twitter.com/JFSIII/status/1804468488783745455)
In that example (nice one, btw, thanks) the issue is with `bazBum` vs `fooBar`.

I made some small tweaks in DO [https://tsplay.dev/w2oG4N](https://t.co/3ptISASm39)

`await stub.fooBar()` works as expected
`await stub.bazBum()` has type never & gives "'await' has no effect ..."

Interesting that `| undefined` fixes

[Mateusz Burzyński](https://twitter.com/AndaristRake) [@AndaristRake](https://twitter.com/AndaristRake) [Jun 23](https://twitter.com/AndaristRake/status/1804786264190673029)
it's nice that today `Serializable` is a union but this creates a problem here, interfaces don't have "inferable indexes" so checking if they extend `Serializable` fails for them (as u already know)

[Mateusz Burzyński](https://twitter.com/AndaristRake) [@AndaristRake](https://twitter.com/AndaristRake) [Jun 23](https://twitter.com/AndaristRake/status/1804786492406989099)
one way to fix this would be to make `Serializable` a conditional type, this was in a fallback branch u could check smth like:

`T[keyof T] extends Serializable ? ... : never`

[Mateusz Burzyński](https://twitter.com/AndaristRake) [@AndaristRake](https://twitter.com/AndaristRake) [Jun 23](https://twitter.com/AndaristRake/status/1804786769847550068)
this would check if its *members* are serializable, not if the whole type is serializable. First u'd have to check if it doesn't have any symbol members as those are not serialiable. But other than that, it should work, i think

[Mateusz Burzyński](https://twitter.com/AndaristRake) [@AndaristRake](https://twitter.com/AndaristRake) [Jun 23](https://twitter.com/AndaristRake/status/1804786813539627143)
And it would be pretty close to what `Serializable` is doing today already

-----

I _think_ https://github.com/cloudflare/workerd/pull/2040 tries the `T[keyof T] extends Serializable ? ... : never` but I'm not entirely sure.

I mainly wanted to bring andarist's comments/suggestions to the relevant place.

Also, we now have TS Playgrounds to use for this!

**Comment by kentonv** [comment]
@penalosa Can you help figure out how we can adjust the types here to solve this?

**Comment by penalosa** [comment]
@kentonv I think the best bet is probably to get https://github.com/cloudflare/workerd/pull/2040 in (I'm updating the tests to make it work). It'll mean that class instances won't throw type errors when passed across an RPC boundary, but they will throw pretty clear runtime errors and so I think that's an acceptable tradeoff for now. If people still run into issues we can completely remove the `never` branch at some point in the future

**Comment by alpertuna** [comment]
Hi @penalosa I completed the fix #2040 and updated the tests. I want to highlight that, even after the fix, there are still cases where some objects cannot be serialized as expected. However, these non-serializable cases are a positive aspect, and the tests cover these scenarios as well.
See the commit: https://github.com/cloudflare/workerd/commit/25cfd1a4ebe1fd33292b0b35fd34ec5215982147

**Comment by alpertuna** [comment]
Reply to @jfsiii https://github.com/cloudflare/workerd/issues/2003#issuecomment-2267516475

> I _think_ #2040 tries the `T[keyof T] extends Serializable ? ... : never` but I'm not entirely sure.

That's true; however, this statement `T[keyof T] extends Serializable ? ... : never` only checks top-level interfaces and not recursive sub-level interfaces. Therefore, it needs to be implemented recursively, which is why the fix became a bit more complex.
